pkg/framework/resource/resource.go

```diff
@@ -76,14 +77,22 @@ func New(config Config) (*Resource, error) {
 	return c, nil
 }
 
-func (r *Resource) Delete(name string) error {
+func (r *Resource) EnsureDeleted(ctx context.Context, name string) error {
```

Comment:
I think that idempotency should just be a general thing we implement everywhere and we can simply leave out the "ensure" part from the naming. We changed the name here now a couple of times and it would be nice to settle with something clean and simple.

Reply:
Agree we should decide on the naming and stick with it. Also agree that code should be idempotent in most cases. So I prefer `Delete` over `EnsureDeleted`.

I think `Install` was to align with Helm but we could use `Create` so we use the CRUD verbs. AIUI we use `Update` explicitly in tests to get resources in the state we need them. So we need to be careful if we change that.

Reply:
If we take the idempotent apprach we need to change error handling also. So changing a name is not an issue here. I'd prefer these things to be explicit. Otherwise it's hard to tell if the function is well handled when reading the code. I still expect `Delete` to return `isNotFoundError`. Let's please stick with that notion to make things clear & explicit.

Question is if the idempotent apprach is right direction here. I'm pretty sure that for deletion it is. But I'm still uncertain about creation/update.

Reply:
I'd be fine with having both `Delete` and `EnsureDeleted` with the ensure method using the delete. Since this is a low level primitive I think that would be useful.

Maybe `EnsureCreated` would be useful too. But we need to keep `Update`. Changing `Install` to `Create` will cause some code churn and not sure its worth it.

Reply:
I like the idea of having function name giving insight on the function behavior. It goes in the same direction as golang `Must*` functions which always panics and return no error.

Reply:
Here is PR https://github.com/giantswarm/fmt/pull/26

Back to my original question. Does it make sense to make this API idempotent? Do we rely on `Update` functionality? 

Maybe it makes sense to have such API:

- EnsureCreated
- EnsureDeleted
- Update

Reply:
I agree with ross, `EnsureDeleted` should use `Delete`. I can easily think of situations in our E2E tests where we need some feedback that something was gone already when we try to `Delete` it. IMO this here is a bit special since we use it in our test cases. 

In general I agree that we should aim for idempotency, but I don't see that harm of having both `EnsureDeleted` and `Delete` here.

Reply:
OK, let's have both then.

Reply:
We use `Update` here https://github.com/giantswarm/chart-operator/blob/master/integration/test/basic/basic_test.go#L91. Not sure if there are other places.

I'm fine with adding the Ensure methods as long as we keep the existing methods.

Reply:
`Delete` restored.